[SLP] NFC. Replace MainOp and AltOp in TreeEntry with InstructionsState.

HanKuanChen opened the pull request:
Add TreeEntry::hasState.
Add assert for getTreeEntry.
Remove the OpValue parameter from the canReuseExtract function.
Remove the Opcode parameter from the ComputeMaxBitWidth lambda function.

HanKuanChen committed:
[SLP] NFC. Replace MainOp and AltOp in TreeEntry with InstructionsState.

InstructionsState will have default constructor.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -836,7 +836,7 @@ class InstructionsState {
   }
 
-  InstructionsState() = delete;
+  InstructionsState() = default;
```
This is the change that should be avoided

HanKuanChen replied:
The problem is the constructor of `TreeEntry`.
```
  struct TreeEntry {
    using VecTreeTy = SmallVector<std::unique_ptr<TreeEntry>, 8>;
    TreeEntry(VecTreeTy &Container) : Container(Container) {}
```
`Container` is just used to dump the SLP tree. See 95da05c3f5064c733dc486a43f262a55deb03dc5.

alexey-bataev replied:
Why do you want to store state at all?

HanKuanChen replied:
1. From `setOperations` we can know it should store InstructionsState.
```
    void setOperations(const InstructionsState &S) {
      MainOp = S.getMainOp();
      AltOp = S.getAltOp();
    }
```
2. The `TreeEntry` reimplemented all of the functions of `InstructionsState`. If we need to add some functions in `InstructionsState`, we may need to reimplement it in `TreeEntry` again.
3. I expect to add some functions in `InstructionsState` to solve the performance issue for https://github.com/llvm/llvm-project/pull/112181.

alexey-bataev replied:
> 1. From `setOperations` we can know it should store InstructionsState.
> 
> ```
>     void setOperations(const InstructionsState &S) {
>       MainOp = S.getMainOp();
>       AltOp = S.getAltOp();
>     }
> ```
> 
> 2. The `TreeEntry` reimplemented all of the functions of `InstructionsState`. If we need to add some functions in `InstructionsState`, we may need to reimplement it in `TreeEntry` again.
> 3. I expect to add some functions in `InstructionsState` to solve the performance issue for [[SLP] Make getSameOpcode support different instructions if they have same semantics. #112181](https://github.com/llvm/llvm-project/pull/112181).

And how does current infrastructure prevent it?

HanKuanChen replied:
The problem is when we call `newTreeEntry`, we don't provide the `MainOp` and `AltOp`.
```
  TreeEntry *newTreeEntry(ArrayRef<Value *> VL,
                          TreeEntry::EntryState EntryState,
                          std::optional<ScheduleData *> Bundle,
                          const InstructionsState &S,
                          const EdgeInfo &UserTreeIdx,
                          ArrayRef<int> ReuseShuffleIndices = {},
                          ArrayRef<unsigned> ReorderIndices = {}) {
    assert(((!Bundle && EntryState == TreeEntry::NeedToGather) ||
            (Bundle && EntryState != TreeEntry::NeedToGather)) &&
           "Need to vectorize gather entry?");
    // Gathered loads still gathered? Do not create entry, use the original one.
    if (GatheredLoadsEntriesFirst.has_value() &&
        EntryState == TreeEntry::NeedToGather &&
        S.getOpcode() == Instruction::Load && UserTreeIdx.EdgeIdx == UINT_MAX &&
        !UserTreeIdx.UserTE)
      return nullptr;
    VectorizableTree.push_back(std::make_unique<TreeEntry>(VectorizableTree));
    TreeEntry *Last = VectorizableTree.back().get();
```
It will make `TreeEntry` call the default constructor of `InstructionsState`. However, it is deleted.

This patch will make `InstructionsState` can be constructed by default constructor.

alexey-bataev replied:
It is provided in `const InstructionsState &S,`

HanKuanChen committed:
Merge remote-tracking branch 'upstream/main' into slp-InstructionsState

HanKuanChen committed:
fix assert which is occurred by InstructionsState

    getMainOp() -> isInstructionsStateValid()
    getMainOp() -> isInstructionsStateValid()? getMainOp(): nullptr
    getOpcode() -> isInstructionsStateValid()
    getOpcode() -> isInstructionsStateValid()? getOpcode(): 0
    getOpcode() -> isInstructionsStateValid() && getOpcode()
 isAltShuffle() -> isInstructionsStateValid() && isAltShuffle()
!isAltShuffle() -> !isInstructionsStateValid() || !isAltShuffle()

HanKuanChen committed:
fix conflict

HanKuanChen committed:
add assert for getTreeEntry

HanKuanChen committed:
OpValue is not used in canReuseExtract

HanKuanChen committed:
add assert for isInstructionsStateValid

HanKuanChen committed:
get opcode from TreeEntry instead of the parameter

HanKuanChen committed:
merge isInstructionsStateValid

HanKuanChen committed:
clang-format

HanKuanChen committed:
add assert for isGather

HanKuanChen committed:
fix isAltShuffle

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -3261,7 +3263,12 @@ class BoUpSLP {
     /// Checks if the current node is a gather node.
-    bool isGather() const {return State == NeedToGather; }
+    bool isGather() const {
+      assert(
+          (State == NeedToGather || S.valid()) &&
+          "InstructionsState must be valid if the TreeEntry is not gathered.");
```
No need for assertion here

HanKuanChen replied:
Why?

alexey-bataev replied:
It just looks strange to have this assertion here

HanKuanChen replied:
I think we should keep it.

alexey-bataev replied:
I understand what you are trying to do here, but we never do this in simple unrelated functions. This function shall check the state of the node, instead we check the state of the InstructionsState. Better to have this assertion in functions, directly related to InstructionsState, not here

HanKuanChen replied:
Got it. The check will be in other PR.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -3402,30 +3403,27 @@ class BoUpSLP {
+    unsigned getAltOpcode() const { return S.getAltOpcode(); }
+
+    bool isInstructionsStateValid() const {
+      assert((S.valid() || isGather()) &&
+             "Invalid InstructionsState must be gathered.");
```
No need for assertion here

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -3402,30 +3403,27 @@ class BoUpSLP {
-      return AltOp ? AltOp->getOpcode() : 0;
+    unsigned getAltOpcode() const { return S.getAltOpcode(); }
+
+    bool isInstructionsStateValid() const {
```
Can you try to invent something shorter? `isValidState`, `hasState`?

HanKuanChen replied:
`isOpStateValid()`?

alexey-bataev replied:
Mybe just `hasState`?

HanKuanChen committed:
rename isInstructionsStateValid to hasState

HanKuanChen committed:
remove assert in isGather and hasState

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -3325,9 +3327,7 @@ class BoUpSLP {
-    /// The main/alternate instruction.
-    Instruction *MainOp = nullptr;
-    Instruction *AltOp = nullptr;
+    InstructionsState S = InstructionsState::invalid();
```
Add a comment for this new data member

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -14846,14 +14843,15 @@ ResTy BoUpSLP::processBuildVector(const TreeEntry *E, Type *ScalarTy,
-    if (!ExtractShuffles.empty() || E->getOpcode() != Instruction::Load ||
-        ((E->getOpcode() == Instruction::Load ||
+    if (!ExtractShuffles.empty() ||
+        (!E->hasState() || E->getOpcode() != Instruction::Load) ||
+        (((E->hasState() && E->getOpcode() == Instruction::Load) ||
```
Drop extra parens

HanKuanChen committed:
apply comment

mikaelholmen commented:
Hi @HanKuanChen 

What about the crash
```clang: /b/1/clang-x86_64-debian-fast/llvm.src/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:821: llvm::Instruction *(anonymous namespace)::InstructionsState::getMainOp() const: Assertion `valid() && "InstructionsState is invalid."' failed.```
seen in the failed build bots?

mikaelholmen commented:
> Hi,
> 
> After recommit of this patch I still see the crash `` clang: ../lib/Transforms/Vectorize/SLPVectorizer.cpp:821: Instruction *(anonymous namespace)::InstructionsState::getMainOp() const: Assertion `valid() && "InstructionsState is invalid."' failed. `` for some downstream tests that https://lab.llvm.org/buildbot/#/builders/56/builds/15893 hit after the first version of this patch. I'll see if I can extract a reproducer but I'm not sure I'll manage before the weekend.

@HanKuanChen :
Reduced opt reproducer:
```opt -passes slp-vectorizer bbi-102823.ll -o /dev/null```
[bbi-102823.ll.gz](https://github.com/user-attachments/files/18378640/bbi-102823.ll.gz)

HanKuanChen commented:
> > Hi,
> > After recommit of this patch I still see the crash `` clang: ../lib/Transforms/Vectorize/SLPVectorizer.cpp:821: Instruction *(anonymous namespace)::InstructionsState::getMainOp() const: Assertion `valid() && "InstructionsState is invalid."' failed. `` for some downstream tests that https://lab.llvm.org/buildbot/#/builders/56/builds/15893 hit after the first version of this patch. I'll see if I can extract a reproducer but I'm not sure I'll manage before the weekend.
> 
> @HanKuanChen : Reduced opt reproducer: `opt -passes slp-vectorizer bbi-102823.ll -o /dev/null` [bbi-102823.ll.gz](https://github.com/user-attachments/files/18378640/bbi-102823.ll.gz)

Really appreciate you report. This is fixed by https://github.com/llvm/llvm-project/pull/122443/commits/5853a032bd078b15c1b7d0232452f129cb5e4c83 now.
